src/sentry/tasks/weekly_reports.py

```diff
@@ -0,0 +1,808 @@
+        self.project = project
+
+        # Array of (group_id, group_history, count)
+        self.key_errors = []
+        # Array of (transaction_name, count_this_week, p95_this_week, count_last_week, p95_last_week)
+        self.key_transactions = []
+        # Array of (Group, count)
+        self.key_performance_issues = []
+
+        # Dictionary of { timestamp: count }
+        self.error_count_by_day = {}
+        # Dictionary of { timestamp: count }
+        self.transaction_count_by_day = {}
+
+    def __repr__(self):
+        return f"{self.key_errors}, Errors: [Accepted {self.accepted_error_count}, Dropped {self.dropped_error_count}]\nTransactions: [Accepted {self.accepted_transaction_count} Dropped {self.dropped_transaction_count}]"
+
+
+# The entry point. This task is scheduled to run every week.
+@instrumented_task(
+    name="sentry.tasks.weekly_reports.schedule_organizations",
+    queue="reports.prepare",
+    max_retries=5,
+    acks_late=True,
+)
+def schedule_organizations(dry_run=False, timestamp=None, duration=None):
+    if timestamp is None:
+        # The time that the report was generated
+        timestamp = to_timestamp(floor_to_utc_day(timezone.now()))
+
+    if duration is None:
+        # The total timespan that the task covers
+        duration = ONE_DAY * 7
+
+    organizations = Organization.objects.filter(status=OrganizationStatus.ACTIVE)
+    for i, organization in enumerate(
+        RangeQuerySetWrapper(organizations, step=10000, result_value_getter=lambda item: item.id)
+    ):
+        if features.has("organizations:weekly-email-refresh", organization):
+            # Create a celery task per organization
+            prepare_organization_report.delay(timestamp, duration, organization.id, dry_run=dry_run)
+
+
+# This task is launched per-organization.
+@instrumented_task(
+    name="sentry.tasks.weekly_reports.prepare_organization_report",
+    queue="reports.prepare",
+    max_retries=5,
+    acks_late=True,
+)
+def prepare_organization_report(
+    timestamp, duration, organization_id, dry_run=False, target_user=None
+):
+    organization = Organization.objects.get(id=organization_id)
+    ctx = OrganizationReportContext(timestamp, duration, organization)
+
+    # Run organization passes
+    with sentry_sdk.start_span(op="weekly_reports.user_project_ownership"):
+        user_project_ownership(ctx)
+    with sentry_sdk.start_span(op="weekly_reports.project_event_counts_for_organization"):
+        project_event_counts_for_organization(ctx)
+    with sentry_sdk.start_span(op="weekly_reports.organization_project_issue_summaries"):
+        organization_project_issue_summaries(ctx)
+
+    with sentry_sdk.start_span(op="weekly_reports.project_passes"):
+        # Run project passes
+        for project in organization.project_set.all():
+            project_key_errors(ctx, project)
+            project_key_transactions(ctx, project)
+            project_key_performance_issues(ctx, project)
```

Comment:
These three focus on querying snuba for the top 3 errors, transactions and performance issues.

I'm concerned on how much of an impact this will have for orgs with hundreds of projects. Once we have this in production we will be able to see what the outliers.

Is there a limitation with Snuba  and using parallelization?

Reply:
For orgs with hundreds of projects, a lot of them are just going to be empty.


There is the [LimitBy](https://getsentry.github.io/snuba-sdk/expressions.html?highlight=limit#snuba_sdk.orderby.LimitBy) [clause](https://clickhouse.com/docs/en/sql-reference/statements/select/limit-by/), but the problem is that the events dataset does not have an "org_id" column. You can also do `group_id in [a long list of group ids]` but it's hard to tell which one is gonna be faster.